src/ledger/backend/ethereumOnChainQuerier.ts

```diff
@@ -0,0 +1,54 @@
+import { providers, Signer } from "ethers";
+
+import { IERC721__factory } from ".";
+import { Tokens } from "../assets";
+
+import { OnChainQuerier } from "../onChainQuerier";
+
+export class EthereumOnChainQuerier implements OnChainQuerier {
+	readonly provider: providers.Provider | Signer;
+
+	constructor(provider: providers.Provider | Signer) {
+		this.provider = provider;
+	}
+
+	async queryTokensOwnedByAddress(
+		token: string,
+		address: string,
+	): Promise<Tokens> {
+		const erc721 = IERC721__factory.connect(token, this.provider);
+		const outboundFilter = erc721.filters.Transfer(address, null, null);
+		const inboundFilter = erc721.filters.Transfer(null, address, null);
+		const [outbound, inbound] = await Promise.all([
+			erc721.queryFilter(outboundFilter),
+			erc721.queryFilter(inboundFilter),
+		]);
+
+		outbound.sort((x, y) => x.blockNumber - y.blockNumber);
+		inbound.sort((x, y) => x.blockNumber - y.blockNumber);
+
+		const tokenOwnedMemo: { [key: string]: boolean } = {};
+
+		while (outbound.length > 0 || inbound.length > 0) {
```

Comment:
This is what happens when you start fixing an algorithm instead of taking a break and starting fresh. :sob: 

I agree, this is easier to understand and equally performant.